addOns/pscan/src/test/java/org/zaproxy/addon/pscan/internal/scanner/PassiveScanControllerUnitTest.java

```diff
@@ -0,0 +1,397 @@
+        given(scanRuleManager.getScanners()).willReturn(List.of(scanner));
+
+        // When
+        psc.start();
+        long testStartTime = System.currentTimeMillis();
+        psc.responseReceived();
+        scanState.waitScanStarted();
+        PassiveScanTask oldestTask = psc.getOldestRunningTask();
+        List<PassiveScanTask> tasks = psc.getRunningTasks();
+        int recordsToScan = psc.getRecordsToScan();
+        scanState.continueScan();
+        scanState.waitScanFinished();
+        long testEndTime = System.currentTimeMillis();
+        sleep(500);
+
+        // Then
+        assertThat(psc.getRecordsToScan(), is(equalTo(0)));
+        assertThat(psc.getOldestRunningTask(), is(nullValue()));
+        assertThat(psc.getRunningTasks().size(), is(equalTo(0)));
+        assertThat(oldestTask.getCurrentScanner().getName(), is(equalTo("TPS")));
+        assertThat(oldestTask.getURI().toString(), is(equalTo(exampleUrl1)));
+        assertThat(oldestTask.getStartTime(), is(greaterThan(testStartTime)));
+        assertThat(testEndTime, is(greaterThanOrEqualTo(oldestTask.getStartTime())));
+        assertThat(tasks.size(), is(equalTo(2)));
+        assertThat(recordsToScan, is(equalTo(2)));
+        assertThat(tasks.get(0).getCurrentScanner().getName(), is(equalTo("TPS")));
+        assertThat(tasks.get(0).getHistoryReference().getHistoryId(), is(equalTo(1)));
+        assertThat(tasks.get(1).getHistoryReference().getHistoryId(), is(equalTo(2)));
+        assertThat(tasks.get(0).getURI().toString(), is(equalTo(exampleUrl1)));
+        assertThat(tasks.get(1).getURI().toString(), is(equalTo(exampleUrl2)));
+    }
+
+    static class TestPassiveScanner implements PassiveScanner {
+
+        private final ScanState scanState;
+        private boolean enabled;
+        private String name;
+
+        TestPassiveScanner(boolean enabled, ScanState scanState) {
+            this("", enabled, scanState);
+        }
+
+        TestPassiveScanner(String name, boolean enabled, ScanState scanState) {
+            this.name = name;
+            this.enabled = enabled;
+            this.scanState = scanState;
+        }
+
+        @Override
+        public void scanHttpRequestSend(HttpMessage msg, int id) {
+            scanState.scanStarted();
+            scanState.holdScan();
+        }
+
+        @Override
+        public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+            scanState.scanFinished();
+        }
+
+        @Override
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public void setEnabled(boolean enabled) {}
+
+        @Override
+        public boolean isEnabled() {
+            return enabled;
+        }
+
+        @Override
+        public boolean appliesToHistoryType(int historyType) {
+            return true;
+        }
+    }
+
+    private static class ScanState {
```

Comment:
Is this the kind of thing you had mentioned should work for the active stuff I looked at 6 or 9 months ago?

Reply:
No, that would be with lock(s).